// @chainflip/rpc/dist/common.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});// src/common.ts



















var _parserscjs = require('./parsers.cjs');

var rpcResult = {
  broker_requestSwapDepositAddress: _parserscjs.brokerRequestSwapDepositAddress,
  cf_account_info: _parserscjs.cfAccountInfo,
  cf_boost_pools_depth: _parserscjs.cfBoostPoolsDepth,
  cf_environment: _parserscjs.cfEnvironment,
  cf_funding_environment: _parserscjs.cfFundingEnvironment,
  cf_ingress_egress_environment: _parserscjs.cfIngressEgressEnvironment,
  cf_pool_orders: _parserscjs.cfPoolOrders,
  cf_pool_price_v2: _parserscjs.cfPoolPriceV2,
  cf_pools_environment: _parserscjs.cfPoolsEnvironment,
  cf_supported_assets: _parserscjs.cfSupportedAsssets,
  cf_swap_rate: _parserscjs.cfSwapRate,
  cf_swap_rate_v2: _parserscjs.cfSwapRateV2,
  cf_swapping_environment: _parserscjs.cfSwappingEnvironment,
  chain_getBlockHash: _parserscjs.chainGetBlockHash,
  cf_boost_pool_details: _parserscjs.cfBoostPoolDetails,
  cf_boost_pool_pending_fees: _parserscjs.cfBoostPoolPendingFees,
  state_getMetadata: _parserscjs.stateGetMetadata,
  state_getRuntimeVersion: _parserscjs.stateGetRuntimeVersion
};



exports.rpcResponse = _parserscjs.rpcResponse; exports.rpcResult = rpcResult;
