// @chainflip/utils/dist/bytes.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/bytes.ts
var bytes_exports = {};
__export(bytes_exports, {
  bytesToHex: () => bytesToHex,
  decodeBytesWithCharset: () => decodeBytesWithCharset,
  encodeBytesWithCharset: () => encodeBytesWithCharset,
  hexToBytes: () => hexToBytes,
  reverseBytes: () => reverseBytes
});
module.exports = __toCommonJS(bytes_exports);

// src/guard.ts
var createIsGuard = (type) => (value) => typeof value === type;
var isString = createIsGuard("string");
var isNumber = createIsGuard("number");
var isBigInt = createIsGuard("bigint");
var isBoolean = createIsGuard("boolean");
var isSymbol = createIsGuard("symbol");
var isObject = createIsGuard("object");
var isUndefined = createIsGuard("undefined");

// src/assertion.ts
function assert(condition, message = "assertion failed", Constructor = Error) {
  if (!condition) {
    throw new Constructor(message);
  }
}

// src/bytes.ts
var bytesToHex = (input) => {
  const bytes = new Uint8Array(input);
  return `0x${Array.from(bytes).map((byte) => byte.toString(16).padStart(2, "0")).join("")}`;
};
var hexToBytes = (input) => {
  assert(/^0x[\da-f]*$/i.test(input) && input.length % 2 === 0, "Invalid hex string");
  const hex = input.slice(2);
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < hex.length; i += 2) {
    bytes[i / 2] = Number.parseInt(hex.slice(i, i + 2), 16);
  }
  return bytes;
};
var convertBase = (bytes, fromBase, toBase) => {
  const result = [];
  for (const byte of bytes) {
    let carry = byte;
    for (let i = 0; i < result.length; i += 1) {
      carry += result[i] * fromBase;
      result[i] = carry % toBase;
      carry = Math.floor(carry / toBase);
    }
    while (carry !== 0) {
      result.push(carry % toBase);
      carry = Math.floor(carry / toBase);
    }
  }
  let leadingZeros = 0;
  while (bytes[leadingZeros] === 0) {
    leadingZeros += 1;
    result.push(0);
  }
  return result.reverse();
};
var encodeBytesWithCharset = (bytes, charset) => convertBase(bytes, 256, charset.length).map((charCode) => charset.charAt(charCode)).join("");
var decodeBytesWithCharset = (input, charset) => {
  assert(new RegExp(`^[${charset}]*$`).test(input), "Invalid input");
  const charMap = Object.fromEntries([...charset].map((char, index) => [char, index]));
  const bytes = input.split("").map((char) => charMap[char]);
  return new Uint8Array(convertBase(bytes, charset.length, 256));
};
var addPrefix = (input) => {
  const [, bytes] = /^(?:0x)?([a-f\d]*)$/i.exec(input) || [];
  assert(bytes, "Invalid hex string");
  return `0x${bytes}`;
};
function reverseBytes(input) {
  const reversed = bytesToHex(hexToBytes(addPrefix(input)).reverse());
  return input.startsWith("0x") ? reversed : reversed.slice(2);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  bytesToHex,
  decodeBytesWithCharset,
  encodeBytesWithCharset,
  hexToBytes,
  reverseBytes
});
